src/node/rpc/frontend.h

```diff
@@ -198,6 +199,173 @@ namespace ccf
+      const ccf::ListenInterfaceID& incoming_interface)
+    {
+      switch (resolver.kind)
+      {
+        case (RedirectionResolutionKind::NodeByRole):
+        {
+          const auto role_it = resolver.target.find("role");
+          const bool primary =
+            role_it == resolver.target.end() || role_it.value() == "primary";
+          if (!primary)
+          {
+            return std::nullopt;
+          }
+
+          const auto interface_it = resolver.target.find("interface");
+          const auto target_interface =
+            (interface_it == resolver.target.end()) ?
+            incoming_interface :
+            interface_it.value().get<std::string>();
+
+          const auto primary_id = consensus->primary();
+          if (!primary_id.has_value())
+          {
+            return std::nullopt;
+          }
+
+          const auto nodes = InternalTablesAccess::get_trusted_nodes(tx);
+          const auto node_it = nodes.find(primary_id.value());
+          if (node_it != nodes.end())
+          {
+            const auto& interfaces = node_it->second.rpc_interfaces;
+
+            const auto target_interface_it = interfaces.find(target_interface);
+            if (target_interface_it != interfaces.end())
+            {
+              return target_interface_it->second.published_address;
+            }
+          }
+          else
+          {
+            return std::nullopt;
+          }
+          break;
+        }
+
+        case (RedirectionResolutionKind::StaticAddress):
+        {
+          return resolver.target["address"].get<std::string>();
+          break;
+        }
+      }
+
+      return std::nullopt;
+    }
+
+    RedirectionResolverConfig get_redirect_resolver_config(
+      endpoints::RedirectionStrategy strategy,
+      const ccf::NodeInfoNetwork_v2::NetInterface::Redirections& redirections)
+    {
+      switch (strategy)
+      {
+        case (ccf::endpoints::RedirectionStrategy::None):
+        {
+          return {};
+        }
+
+        case (ccf::endpoints::RedirectionStrategy::ToPrimary):
+        {
+          return redirections.to_primary;
+        }
+      }
+    }
+
+    bool check_redirect(
+      kv::ReadOnlyTx& tx,
+      std::shared_ptr<ccf::RpcContextImpl> ctx,
+      const endpoints::EndpointDefinitionPtr& endpoint,
+      const ccf::NodeInfoNetwork_v2::NetInterface::Redirections& redirections)
+    {
+      auto rs = endpoint->properties.redirection_strategy;
+
+      switch (rs)
+      {
+        case (ccf::endpoints::RedirectionStrategy::None):
+        {
+          return false;
+        }
+
+        case (ccf::endpoints::RedirectionStrategy::ToPrimary):
+        {
+          const bool is_primary =
+            (consensus != nullptr) && consensus->can_replicate();
+
+          // Note: This check is included for parity with forwarding. If we
+          // should redirect, but can't for fundamental early-node-lifecycle
+          // reasons, should we try to execute it locally?
```

Comment:
I don't think we should try to execute locally, the mixed forward/redirect situation is sufficiently complicated as it is.

Reply:
Agreed. Tracking this as the second follow-up TODO here: https://github.com/microsoft/CCF/issues/6082

I expect we can remove this, but will need to plumb through some minor test changes to do so.